crates/client/src/lib.rs

```diff
@@ -797,75 +872,116 @@ impl<R: RegistryStorage, C: ContentStorage, N: NamespaceMapStorage> Client<R, C,
         self.registry
-            .store_operator(self.api.get_warg_registry(), operator)
+            .store_operator(registry_domain, operator)
             .await?;
 
         for package in packages.values_mut() {
-            package.checkpoint = Some(checkpoint.clone());
+            package.registry = registry_domain
+                .cloned()
+                .or_else(|| Some(self.url().registry_domain()));
+            package.checkpoint = Some(checkpoint.clone()); // updated to this checkpoint
             self.registry
-                .store_package(self.api.get_warg_registry(), package)
+                .store_package(registry_domain, package)
                 .await?;
         }
 
         self.registry
-            .store_checkpoint(self.api.get_warg_registry(), ts_checkpoint)
+            .store_checkpoint(registry_domain, &ts_checkpoint)
             .await?;
 
-        Ok(())
+        // return packages to be retrieved from other registries
+        Ok(federated_packages)
     }
 
+    /// Update checkpoint for list of packages
     async fn update_checkpoints<'a>(
-        &mut self,
-        ts_checkpoints: IndexMap<std::string::String, SerdeEnvelope<TimestampedCheckpoint>>,
-        mut packages: IndexMap<String, Vec<PackageInfo>>,
+        &self,
+        packages: impl IntoIterator<Item = &mut PackageInfo>,
     ) -> Result<(), ClientError> {
-        for (name, ts_checkpoint) in ts_checkpoints {
-            if self.url().safe_label() != name {
-                self.api
-                    .set_warg_registry(Some(RegistryDomain::from_str(&name)?));
+        // first collect the packages that we already have namespace mappings for
+        let mut federated_packages: IndexMap<Option<RegistryDomain>, Vec<&mut PackageInfo>> =
+            IndexMap::new();
+        for package in packages.into_iter() {
+            let registry_domain = self.get_warg_registry(package.name.namespace()).await?;
+            if let Some(package_set) = federated_packages.get_mut(&registry_domain) {
+                package_set.push(package);
             } else {
-                self.api.set_warg_registry(None)
+                federated_packages.insert(registry_domain, vec![package]);
             }
-            let mut packages = packages.get_mut(&name.clone());
-            if let Some(pkgs) = &mut packages {
-                self.update_checkpoint(&ts_checkpoint, pkgs.as_mut_slice())
-                    .await?;
+        }
+
+        while let Some((registry_domain, packages)) = federated_packages.pop() {
+            for (registry_domain, mut packages) in self
+                .update_packages_and_return_federated_packages(registry_domain.as_ref(), packages)
+                .await?
+                .into_iter()
+            {
+                if let Some(package_set) = federated_packages.get_mut(&registry_domain) {
+                    package_set.reserve(packages.len());
+                    package_set.append(&mut packages);
```

Comment:
I'm pretty sure you just want `extend`:
```suggestion
                    package_set.extend(packages);
```

Reply:
Fixed